backend/src/main/java/com/redhat/cloud/notifications/db/EventResources.java

```diff
@@ -30,27 +31,67 @@
+                .onItem().transformToUni(eventIds -> {
+                    String hql = "SELECT DISTINCT e FROM Event e " +
+                            "JOIN FETCH e.eventType et JOIN FETCH et.application a JOIN FETCH a.bundle b " +
+                            "LEFT JOIN FETCH e.historyEntries he LEFT JOIN FETCH he.endpoint " +
+                            "WHERE e.accountId = :accountId AND e.id IN (:eventIds)";
+
+                    if (orderByCondition.isPresent()) {
+                        hql += orderByCondition.get();
+                    }
+
+                    return session.createQuery(hql, Event.class)
+                            .setParameter("accountId", accountId)
+                            .setParameter("eventIds", eventIds)
+                            .getResultList();
+                });
+    }
+
+    public Uni<Long> count(String accountId, Set<UUID> bundleIds, Set<UUID> appIds, String eventTypeDisplayName,
+                      LocalDate startDate, LocalDate endDate) {
+        String hql = "SELECT COUNT(*) FROM Event e JOIN e.eventType et JOIN et.application a JOIN a.bundle b " +
                 "WHERE e.accountId = :accountId";
 
         hql = addHqlConditions(hql, bundleIds, appIds, eventTypeDisplayName, startDate, endDate);
 
+        Mutiny.Query<Long> query = session.createQuery(hql, Long.class);
+        setQueryParams(query, accountId, bundleIds, appIds, eventTypeDisplayName, startDate, endDate);
+
+        return query.getSingleResult();
+    }
+
+    private Optional<String> getOrderByCondition(String sortBy) {
         if (sortBy == null) {
-            hql += " ORDER BY e.created DESC";
+            return Optional.of(" ORDER BY e.created DESC");
         } else {
             Matcher sortByMatcher = SORT_BY_PATTERN.matcher(sortBy);
             if (sortByMatcher.matches()) {
                 String sortField = getSortField(sortByMatcher.group(1));
                 String sortDirection = sortByMatcher.group(2);
-                hql += " ORDER BY " + sortField + " " + sortDirection + ", e.created DESC";
+                return Optional.of(" ORDER BY " + sortField + " " + sortDirection + ", e.created DESC");
+            } else {
+                return Optional.empty();
             }
         }
+    }
+
+    private Uni<List<UUID>> getEventIds(String accountId, Set<UUID> bundleIds, Set<UUID> appIds, String eventTypeDisplayName,
+                                        LocalDate startDate, LocalDate endDate, Integer limit, Integer offset, Optional<String> orderByCondition) {
+        String hql = "SELECT e.id FROM Event e JOIN e.eventType et JOIN et.application a JOIN a.bundle b " +
+                "WHERE e.accountId = :accountId";
+
+        hql = addHqlConditions(hql, bundleIds, appIds, eventTypeDisplayName, startDate, endDate);
 
-        Mutiny.Query<Event> query = session.createQuery(hql, Event.class);
+        if (orderByCondition.isPresent()) {
+            hql += orderByCondition.get();
```

Comment:
Do we really need the order by here? 
I think it's something that we don't really care about in the list of ids.

Reply:
Yes, we have to order the records because of the limit.

Reply:
From the [Postgres doc](https://www.postgresql.org/docs/13/queries-limit.html):

> When using LIMIT, it is important to use an ORDER BY clause that constrains the result rows into a unique order. Otherwise you will get an unpredictable subset of the query's rows. You might be asking for the tenth through twentieth rows, but tenth through twentieth in what ordering? The ordering is unknown, unless you specified ORDER BY.

Reply:
I missed the limit :facepalm: }
nothing to see here :grin: